developers_chamber/qa/base.py

```diff
@@ -0,0 +1,131 @@
+
+    def _get_diffs(self):
+        """
+        Returns Git diffs against default branch.
+        """
+        repo = self._get_repo()
+        target_branch = getattr(repo.remotes.origin.refs, self._get_default_branch())
+        target_commit = repo.merge_base(repo.active_branch, target_branch)[0]
+        return target_commit.diff(repo.active_branch.name, create_patch=True)
+
+    def _get_unstaged(self):
+        """
+        Returns unstaged files in the repo.
+        """
+        return self._get_repo().index.diff(None, create_patch=True)
+
+    def _run_command(self, command):
+        """
+        Runs shell command and returns its output.
+        """
+        try:
+            return subprocess.check_output(command, shell=True).decode().strip()
+        except subprocess.CalledProcessError as ex:
+            raise QAError(str(ex), ex.output.decode())
+
+    def _run_check(self):
+        """
+        This function should implement the actual check and should raise `QAError` if failed.
+        """
+        raise NotImplementedError()
+
+    def _cleanup(self):
+        """
+        Cleans up the repo to be fresh for another check.
+        """
+        self._get_repo().git.stash('save')
+
+    def run(self):
+        """
+        Runs the check and cleanup methods.
+        """
+        try:
+            self._run_check()
+        finally:
+            self._cleanup()
+
+
+class QACheckRunner:
+    """
+    Runs multiple checks and evaluates the results while printing a nice output.
+    """
+    LINE_LENGTH = 60
+
+    def __init__(self, *checks):
+        """
+        Arguments:
+            checks: Instances of QACheck class.
+        """
+        self.checks = checks
+        self.results = []
+        self.success = True
+
+    def _print_line(self):
+        LOGGER.info('-' * self.LINE_LENGTH)
+
+    def _print_title(self, title):
+        self._print_line()
+        LOGGER.info(click.style(title, fg='blue', bold=True))
+        self._print_line()
+
+    def _run_checks(self):
+        self._print_title('QA - running checks')
+        for i, check in enumerate(self.checks):
+            LOGGER.info(click.style('[{}/{}] {}...'.format(i + 1, len(self.checks), check.name), fg='blue'))
+            try:
+                check.run()
+            except QAError as ex:
+                self.results.append(click.style('{}\n{}'.format(click.style(str(ex), fg='red'), ex.output)))
+                self.success = False
+            else:
+                self.results.append(click.style(' OK ', bg='green'))
+
+    def _print_results(self):
+        self._print_title('QA - results')
+        for i, result in enumerate(self.results):
+            LOGGER.info(click.style('#{} {}'.format(i + 1, self.checks[i].name), fg='blue'))
+            LOGGER.info(result)
+
+    def run(self):
+        self._run_checks()
+        self._print_results()
+        exit(0 if self.success else 1)
```

Comment:
I still think it'd be nicer to use Click exceptions. However, if both @matllubos and @Formulka are fine with this, I'm fine with it too.

Reply:
IMO this is more explicit in a way I control both success and failure exit code. What is better on using Click exception?

Reply:
We use them everywhere and I don't see any added value in your approach except for it being inconsistent with the rest of the project. Sorry.

Reply:
I do now like use exit function in class which can be used separately. You limit future improvements. It will be better here return for example True/false or some sofisticated response (for example exception), but return exit is weird. It should be used only inside click function

Reply:
OK, that sorta makes sense.

Reply:
TBH I don't like the ClickException solution much - it requires message (i don't want to supply any message here) and it does not use logger to print error (which is therefore printed like normal output). But I will use it, apparently, you insist.